### ISSUE TITLE
Any plans to add repackaging?

### ISSUE THREAD

**Issue by Kein** [healingbrew/UAssetParser#1]
From what I see it should be relatively possible. You need to adjust offsets if you change string values but the rest are free. The only concern arises when Exports length is more than 1 but I never seen DT or ST like that yet

**Comment by healingbrew** [comment]
No plans to add repackaging, mostly because there are some types that are more or less hacked together and may lose some data in the process of parsing.

Some games are real funky with how much they nest structures, and this project loses track of when a structure starts and when it ends. Animation graphs, for example are horrendeous when decoded with this.

**Comment by Kein** [comment]
What metadata are you talking about? I've tested the parser on Bloodstained and from what I see all metadata in its DTables is parsed properly, nothing untouched is left. Are you talking about output JSON? Well yeah, if you dont dump it with JSON it is kinda "lost". The only issue I found is some ints are bitshifted and became multiply of 256.
If you dump FilePackageSummary data it should be possible to pack int into binary, using as a base, but then the issue arises when you change values that would change offset like, add an new item or change string.

**Comment by healingbrew** [comment]
Interesting, I'll look into it some day but no definite timeframe.

**Comment by Kein** [comment]
There is a 4byte value right after the string termination byte of the last entry in the Names collection. It is not part of the following Imports data but it is present and different for each Data Table. Since itis not relevant to Deserialization you ignore it, but do you recall, by any, chance what is it for?

**Comment by Kein** [comment]
Hey, do you remember why do you set size for FName to 8 and read only 4 bytes for INT32 value and skip the rest 4? I've noticed that some UAsset files have some data in these 4 bytes and it is not always 4th zero bytes.

**Comment by healingbrew** [comment]
I don't quite remember but I think it's some internal type mapping. Not sure.

**Comment by Kein** [comment]
Did you never figure out how to determine if ByteProperty for UArrays is anactual byte or FName? In your case, your EnumenatorVisitor you just read amount of bytes based on ArraySize and then skips to the end, what I'm seeing in my case of UASSET:
![image](https://user-images.githubusercontent.com/623756/62820529-0d925000-bb6e-11e9-8b8f-d1f6547b5e7b.png)

Count is 2, FPropertySize is 20 and the content of the UArray is seems to be 2 FNames, not bytes.  Normally for ByteProperty you check Enum Name but there is none for the UArray.

**Comment by healingbrew** [comment]
The only way to determine that is to get the actual model from the game files.

I suppose you could check it by checking if Size == Count, or Size / 4 == Count.

This is caused by EnumAsByte, which uses FNames instead of bytes but is listed as a ByteProperty in UObjects.

**Comment by Kein** [comment]
I went through all of the DataTables for the game I'm using as an example and there never was a case when ByteProperty inside an UArray was an actual byte. Judging by what you said I assume it wont be then since their model is using FNames?

> game binary/source
You mean the PE executable binary?

**Comment by healingbrew** [comment]
Yes.

And I've not found any actual byte uses as well, but that's probably because nobody uses bytes (and I think the engine prefers full ints anyway)

**Comment by Kein** [comment]
Yeah, that makes sense. On a follow-up note - have you ever figured out what is the has for TextProperty? I know that the Int64 is some serial offset in stream, but the Has puzzles me. Iti s 32 bytes long and does not looks like anything of the standard algos and I dont see mention of any hashing in the TextProperty in UE4 sources

**Comment by healingbrew** [comment]
I think it's 2 hashes.

TextProperty embeds a FText structure, which is what FNames are based on.

1st half = NonCasePreservingHash
2nd half = CasePreservingHash.

Case Preserving Hash is Crc32 (PKZip variant, 0x04C11DB7) & 0xFFFF.
Non-Case Preserving Hash is same as Case preserving but full uppercase.

However there's a difference.

Case-preserving hash works in 32-bit words when calculating the has.
```
crc = (crc >> 8) ^ table[0][(crc ^ char) & 0xFF];
char >>= 8;
crc = (crc >> 8) ^ table[0][(crc ^ char) & 0xFF];
char >>= 8;
crc = (crc >> 8) ^ table[0][(crc ^ char) & 0xFF];
char >>= 8;
crc = (crc >> 8) ^ table[0][(crc ^ char) & 0xFF];
```
Non-case preserving has works with single bytes.
`((crc >> 8) & 0x00FFFFFF) ^ table[(crc ^ char) & 0x000000FF]`

**Comment by healingbrew** [comment]
I think it's 2 hashes.

TextProperty embeds a FText structure, which is what FNames are based on.

1st half = NonCasePreservingHash.
2nd half = CasePreservingHash.

Both are shorts.

Case Preserving Hash is Crc32 (PKZip variant, 0x04C11DB7) & 0xFFFF.
Non-Case Preserving Hash is same as Case preserving but full uppercase.

However there's a difference.

Case-preserving hash works in 32-bit words when calculating the hash.
Non-case preserving hash works with single bytes.

**Comment by Kein** [comment]
No, that's hashes for FNames and FNameEntries, I'm talking about TextProperty. It uses very different Hash mechanism that is nothing like FName's:
https://github.com/healingbrew/UAssetParser/blob/master/UAssetParser/Objects/Visitors/Unreal/UTextProperty.cs

32byte long, while NCP hash and CP hashes are 2bytes ushorts.

**Comment by healingbrew** [comment]
Hash might be the key used for FSet/FMap, not sure.

**Comment by Kein** [comment]
Yeah, this concerns me when comes to editing and serializing back. The Fname hashes are easy ones, but TextPropery hash and FName instance number is the biggest hurdle for serialization back to binary.

**Comment by healingbrew** [comment]
The thing is that it might just be a value used specifically for hash mapping, so it doesn't get validated.

Maybe?

**Comment by Kein** [comment]
I know that changes to TextProperty are reflected in game since I've tested it. Thing is, since it is saved it should have use in some subtle cases. LIke, or example, I had to reindex FName entries after addition or removal but because C# default ORdinalIC string sorting is different than C++ (due to different weight tables) some string were placed differently (Like PB_test_Data and PBTestData) and this apparently breaks stuff in very rare case when UE performs binary search (??? why tho you have index). Also, FNames have damn instance number that can fuck up asset completely.
It just you never know how important can be the data so it always better be safe than sorry.

**Comment by Kein** [comment]
Hey, do you remember by any chance what is the bunch of FF flags at the end of UASSETs which are complemented by UEXPs? Normally, for a plain UASSET it would be just 4 zero bytes padding and then raw serial data but when this data is split into UEXP, the ending of UASSET has some nonsense. I thought it was some hash but the size can differ from 40 to 80 bytes.
![image](https://user-images.githubusercontent.com/623756/71638841-f4548200-2c7b-11ea-96d2-f5037728b763.png)

**Comment by healingbrew** [comment]
They're negative 32-bit numbers, I'm guessing they're table IDs? Maybe to define root objects?

**Comment by Kein** [comment]
Party guessed they seems to be preload-dependencies but couldnt find jack shit in Unreal UObject code regarding that.

**Comment by healingbrew** [comment]
Most likely something specific to blueprints, or something higher level. UObject is just a serialization format, it doesn't really do any computation in runtime.

**Comment by Kein** [comment]
Do you by any chance remember why FCustomVersion elements were chosen to be 16bytes long GUID?
https://github.com/healingbrew/UAssetParser/blob/master/UAssetParser/Formats/FCustomVersion.cs

I'm looking at the header v7 of the FF7 Remake for the 3D meshes and there are 6 entries in CustomVersion and total length is 120 bytes so I'm assuming they are 20 bytes each. Does not look like GUID at all.

**Comment by healingbrew** [comment]
https://docs.unrealengine.com/en-US/API/Runtime/Core/Serialization/FCustomVersion/index.html
It's actually FGuid Key + int32 Version, I never actually found a package with an FCustomVersion so I never actually looked it up properly.

**Comment by Kein** [comment]
Yeah, but that does not look right. I did some browsing of 
https://github.com/EpicGames/UnrealEngine/blob/6c20d9831a968ad3cb156442bebb41a883e62152/Engine/Source/Runtime/Core/Private/Serialization/CustomVersion.cpp
and
https://github.com/EpicGames/UnrealEngine/blob/6c20d9831a968ad3cb156442bebb41a883e62152/Engine/Source/Runtime/Core/Public/Serialization/CustomVersion.h

However, the data does not really match.

```hex
0000000006000000 3F74FCCF8044B043
DF14919373201D17 1E000000ED68B0E4
E94294F40BDA31A2 41BB462E14000000
004A8AD7974658E8 B519A8BAB4467D48
0C0000009F8BF812 FC4A75880CD97CA6
29BD3A3816000000 86F879551F4C3A93
7B08BA832FB96163 0100000032F4552D
8F44997A44E0428B 8C57CE4F00000000
F14600000---------- ---------------------
```
As you can see, count is 6 and then there is 120 bytes. GUID + Int32 pair dont really work here, there are no valid guids that way. For example, first one is: 
`3F74FCCF8044B043DF14919373201D17`
Regardless of endiannes it is not a valid GUID. Something is off.

**Comment by healingbrew** [comment]
GUID is 16 bytes,  Int32 is 4 bytes.
3F74FCCF8044B043DF14919373201D17 is 10 bytes
1E000000 is 4 bytes
The only issue is the last int32, which may be ReferenceCount or a pointer to an FName.

**Comment by Kein** [comment]
My bad, fren, the tool I was using is broken and failed to parse GUID properly. Spent a hour trying to understand why `3F74FCCF8044B043DF14919373201D17` is a valid when the tool wasnt able to parse it.

**Comment by Kein** [comment]
Posting small update here so I can later reference it. I did more research and still came to the same conclusion the GUID is some internal UE one. For example, these are guids I have extracted from the header:
{CFFC743F-4480-43B0-DF14-919373201D17} (30)
{E4B068ED-42E9-F494-0BDA-31A241BB462E} (20)
{D78A4A00-4697-E858-B519-A8BAB4467D48} (12)
{12F88B9F-4AFC-8875-0CD9-7CA629BD3A38} (22)
{5579F886-4C1F-933A-7B08-BA832FB96163} (1) 

the first hex-digit on the CCCC part represents version of the UUID/GUID, it has to be from 1 to 5, according to specs (https://tools.ietf.org/html/rfc4122#section-4.4, this spec also valid for other versions like MS one), however, some of these GUIDs regardless of endianess have invalid versions like 9 or 15. Same for Variant - while first GUID has varian "D" which stands for Microsoft reserved one, the rest often use non-valid reserved for future variants.

**Comment by Kein** [comment]
**Forgot to post my research/testing summary on re-serialization:**

- The one and **ONLY** lossy aspect of this process is instance number tracking for FNames. From my understanding they are tracked globally across the project so we cant inherit **full** data from a particular asset (but we still must inherit what we have and **ALWAYS preserve existing instance numbers**). Editing/adding these must be limited and with a warning. 

- Due to the nature of the object, JSON->Binary isnt really feasible, because we lose too much type information and it is PITA to write custom parser and try to figure out the typing. In my opinion the only reasonable approach is GUI and direct editing the processed object because we have all the types (this way we automatically inherit from the original file)

- The following header entries needs to be adjusted/recalculated:
> All the offsets and the respective counts. That means we can no longer skip any other entries that seemed irrelevant (like PreloadDependency, etc)
> FNames needs to be sorted alphabetically, duplicates or unreferenced entries removed. Default .NET's sorting mechanism with `StringComparer.OrdinalIgnoreCase` works however, it processes things like `ABC_DEF` differently in C++ (or may be it is a custom UE sorter) so you end up with some inconsistencies here. While this is not an issue for the assets since it relies on indexes there might be a case where UE internally uses BinarySearch and it can fail (but I never had such case)

The rest is pretty straightforward

**Comment by Kein** [comment]
Looks like there are breaking changes in 4.23:
https://www.unknowncheats.me/forum/unreal-engine-4-a/366381-unreal-4-23-a.html
https://www.unknowncheats.me/forum/unreal-engine-4-a/387115-unreal-4-23-gnames-fnamepool-enumeration.html

**Comment by healingbrew** [comment]
The core branch is built around serializing, including storing type info in the JSON outputs.
This does mean that the output is WAY more bloated than the netfx flavor. 

There are still some issues like the aformentioned name pool however, I think we should avoid mutating it since it's going to be baked into the binaries for a lot of games soon.

The idea is to not touch the header at all beyond ObjectExport and SerialOffset/SerialSize and bake everything in respect to the vanilla header.

Serializing is incomplete in the core branch, I'll finish implementing it when I can but right now I'm still in the process of rewriting the old property models to the new Core method (which has almost no reflection and is faster, but requires a lot more manual code.)
You can track progress in the core branch: https://github.com/healingbrew/UAssetParser/tree/core

**Comment by Kein** [comment]
Skipping name pool modification means we cant alter the FNames or add/remove them to add new entries. On a bright side it means we don't need to care much about instance number tracking and trying to guess it across whole project in some magic way (doubt it is even doable).

>  WAY more bloated than the netfx flavor. (netfx vs core)

What if we reuse original asset for the structure, thus allowing us to slim down JSON output? That's how I did it originally for Bloodstained but I then realized it would be much simpler to implement GUI front and edit in-memory.

**Comment by healingbrew** [comment]
> What if you reuse original asset for the structure, thus allowing us to slim down JSON output? That's how I did it originally for Bloodstained but I then realized it would be much simpler to implement GUI front and edit in-memory.

Is a possibility, but for the sake of serial conversion it's still easier to store this info. I need to find a game that has a baked FNamePool so I can see if I can use both the baked pool and the embedded pool.

**Comment by Kein** [comment]
> Is a possibility, but for the sake of serial conversion it's still easier to store this info.  

Yeah, it will be super-useful when there are StructObjects or UArrays we would want to modify, however, you also gonna need to deserialize and store ALL the asset data in JSON, including meta, in this case. Or, otherwise, you'd have to fallback to original binary for all the meta (and meta should not be modifiable to begin with).

**Comment by Kein** [comment]
Is DragonLib outdated? I'm referencing assembly build from latest master and there is no `GenericListConverterFactory()`

**Comment by healingbrew** [comment]
The package is in GitHub packages, but I just published it to Nuget. Should have it available there soon.

**Comment by Kein** [comment]
There is probably a need to pass down some meta from Array/Struct wrappers, for example, I know about few objects that change its state depending if they are single or packed into array:

**ByteProperty**
Will be either just a raw byte or plain FName when in Array. The only way to know that from inside the asset is to calculate size of the array: https://github.com/healingbrew/UAssetParser/blob/d8ef124e10510ad65af9720330446e82c513592c/UAssetParser/Objects/Visitors/Unreal/UArrayProperty.cs#L48
That means it needs to know if it is inside array or not to properly serialize back.
_Question_: Assuming original asset has a few entries, each of the entries has a ByteArray and some of these entries have ByteArray empty. How would you figure out which type of ByteArray it is in this case, to read back?  There will be no meta.

**EnumProperty, NameProperty**
Both are plain FNames (without GUID data, etc) when in array.

**TextProperty**
Simplified version of original without the tag, only StreamOffset, GUID, Hash and Value.

**SoftObjectProperty**
Only Package as FName and Path as string

**Comment by healingbrew** [comment]
> 
> 
>     1. There is a separate editor/helper that parses the json and makes editing easier
> 
>     2. or the Parser made with direct editing support from the ground-up (no json, straight in-memory edit and serialization, most likely gonna require gui).

Probably 2. I added a switch to generate typeless information to keep readable JSON output, but this will fail to re-serialize to managed C# objects.

**Comment by Kein** [comment]
https://store.steampowered.com/app/1137350/Filament/

Seems to be using UE 4.24 but so far I havent found anything about FNames pool

**Comment by healingbrew** [comment]
> 
> 
> https://store.steampowered.com/app/1137350/Filament/
> 
> Seems to be using UE 4.24 but so far I havent found anything about FNames pool but the DataTables have some difference in structure:
> [DT_Dialogue.zip](https://github.com/healingbrew/UAssetParser/files/4566784/DT_Dialogue.zip)

FNamePool is embedded in the UAsset, also GUIDs are 32-byte strings and not a 16-byte array like I thought. I don't think this requires much editing to actually make function.

**Comment by Kein** [comment]
Are you talking about Name entries in the header?

**Comment by healingbrew** [comment]
The DT you linked parses fine with the edits I just pushed to Core, implying that the GName/FNamePool isn't used. Maybe it's used for the null entries in the arrays?

**Comment by Kein** [comment]
No idea, I just assumed everything starting from 4.23 automatically cooks with Fname pools as the thread on UC forums states. It is the only new big change I've encountered in last 2 years related to assets

**Comment by Kein** [comment]
Here are 2 examples of TextProperty that do not work for your conditions, one of them has `ReservedHashInt` and `ReservedValueInt` both 0 but has both **Hash** and **Text** and another one has ReservedHashInt` as 0 and `ReservedValueInt` as  `FF 8B 2A 00` and yet has no values for **Hash** and **Text**

You probably could add these 2 ass well into the condition check but then new game can have their own values for `ReservedHashInt` and `ReservedValueInt` and so on, you cant add these forever.

[DT_DialogueSpeakerAttributes.zip](https://github.com/healingbrew/UAssetParser/files/4567113/DT_DialogueSpeakerAttributes.zip)

**Comment by healingbrew** [comment]
Looking at the actual implementation of UTextProperty, it seems to be all sorts of weird. UTextProperty is a FPropertyNode pointing to a FText object.
None of the values have a 64-bit StreamOffset value, but instead have a unsigned 32-bit Flags value.

**Comment by Kein** [comment]
The asset files that were posted in the other issue:
https://github.com/healingbrew/UAssetParser/files/4565857/Problematic_UAsset_Files.zip
Even weirder. After the size and index of property tag is read, it ends with 10 bytes - first 5 are zero bytes, then 0xFF then 4 zero bytes (???)

**Comment by healingbrew** [comment]
the current implementation parses all problem files and known "sane" files.

**Comment by Kein** [comment]
I think in JSON output we should make sure JSON writer make clear separation between integers and float values (`0` vs `0.0`).

Okay, so, for serialization, if you decide to stick with JSON **without** reusing original asset file, here are issues we gonna face:

- Most of the AssetSummary data/header needs to be dumped, except these which we can and will have to recalculate: _BulkDataStartOffset, SerialSize, NameCount, ImportOffset, ExportOffset, DependsOffset, TotalHeaderSize, Generations.NameCount, AssetRegistryDataOffset, PreloadDependencyOffset._ 
- We also need to print InstanceNum for EVERY FName into JSON. Instance Numbers are mega-important, if it has 2 and you put 0 - it can crash the runtime player. They need to be preserved as it is. From my experience with Bloodstained, instance numbers editing in **values** (when value is NameProperty or Enum or Byte as Enum) seems to be more or less fine as they are 0 anyway in 90% cases. But editing FNames and their instances when they are **object keys** often results in issues. In my case I've just appended them to the FName string, like "Name_(int)" and when reading back, if there was no instance number specified (i.e. it didnt end with (int) I just assumed it is 0)
- Reading the values from JSON also adds extra challenge in a form that you first need to rebuild the collection of EVERY FName entry and then sort them and then, rebuild indexes for every **key** and **value** that use FName.
- I dont know if UE uses default c++ string sorting or not, but if the goal is to be as pure as possible in serialization then sorting strings by `OrdinalIgnoreCase` does the job fine, except stuff like `_` has different weight in C# vs C++ so these strings will have different places. This can be redeemed by writing custom sorter but I never got around

The rest of the issues are just type knowledge issues, if you dump all type data it will be fine, but there are still edge cases like I mentioned before: https://github.com/healingbrew/UAssetParser/issues/1#issuecomment-622127966 and I have no idea how to handle them (probably leave to the user to deal with issue if he uses wrong type but this requires the rest of serialization to be very robust so at least user can deduce what went wrong)

**Comment by Kein** [comment]
I'm also not sure if it even make sense for properties to inherit from `AbstractGuidProperty` ever because any of these properties can be inside struct or array where they are mostly stripped down to base value. Original parser had completely separate logic for that.

**Comment by Kein** [comment]
Also, Map stuff:
https://www.gildor.org/smf/index.php/topic,7063.msg35982.html#msg35982

**Comment by healingbrew** [comment]
> I'm also not sure if it even make sense for properties to inherit from `AbstractGuidProperty` ever because any of these properties can be inside struct or array where they are mostly stripped down to base value. Original parser had completely separate logic for that.

Design decision to unify implementations.

> Or may be, since we have custom serialization methods for each type anyway - we ALWAYS inherit from AbstractGuidProperty and just do custom branch logic inside [de]serialization body of the property, i.f. if SerializationMode is array we just skip base type deserialization and go straight to value read?

This is what happens currently. AbstractGuidProperty overrides `Deserialize(Span<byte>, AssetFile, ByRef int)` which itself is called by AbstractProperty's `Deserialize(Span<byte>, AssetFile, ByRef int, SerializationMode)` when `mode == Normal`

Due to the inconsistency of where the Guid is stored, it might be better to just ditch the subclass and implement it on a per-property basis.

**Comment by Kein** [comment]
> This is what happens currently. AbstractGuidProperty overrides `Deserialize(Span<byte>, AssetFile, ByRef int)` which itself is called by AbstractProperty's `Deserialize(Span<byte>, AssetFile, ByRef int, SerializationMode)` when `mode == Normal`

Yeah, I was checking your Array and struct property serialization and noticed this also:
https://github.com/healingbrew/UAssetParser/blob/070db305cf9c6932943156116fd10d5c8cac5783/UObject/Properties/ArrayProperty.cs#L36

Looks like you've completely skip the StructProperty with a tag when it is packed int Array. From my observation on a few assets it seems like StructName matches the `Array name` so the tag could be reconstructed (altho no guarantee some assets wont use different names) but you still skip Struct-related meta data. Any particular reason for that?

**Comment by healingbrew** [comment]
Because we don't have a serialization track for carrying array or map info. However in a recent commit the structdata is kept.

**Comment by Kein** [comment]
I'm testing simple custom DataTable on UE 4.24.3 with TextProperty and it seems like they've changed it again - current logic fails. If I enable `localization` for property there are 3 string now, if I disable it then there is only one string that is deserialized as `Hash`

**Comment by healingbrew** [comment]
> 
> 
> I'm testing simple custom DataTable on UE 4.24.3 with TextProperty and it seems like they've changed it again - current logic fails. If I enable `localization` for property there are 3 string now, if I disable it then there is only one string that is deserialized as `Hash`

There's something with the Flags value that determines this, but I don't have enough data to figure it out.

**Comment by Kein** [comment]
Just gonna drop some observations so far for reference later (4.24):
- Localization _disabled_, Text field populated:
    `02 00 00 00 FF 01 00 00 00`
- Localization _enabled_, Text field populated, Namespace deleted:
    `00 00 00 00 00 01 00 00 00 00`
- Localization enabled, Text field populated, Namespace present, Key(Hash) present:
    `00 00 00 00 00`
- Localization _enabled_, text and namespace deleted, key auto-deleted by UE:
    `00 00 00 00 FF 00 00 00 00`
- Localization _disabled_, text and namespace deleted, key auto-deleted by UE:
    `02 00 00 00 FF 00 00 00 00`

**Comment by Kein** [comment]
Alright, I made TextProperty work, the only thing I could not text is 4.18 version in Arrays/Structs because I dont want to download and install 5GB more of bloat.

**Comment by Kein** [comment]
May be instead of bunch of branching for the TextProperty, instead, implement all revisional properties as separate assembly modules? So during runtime it will load either user-specified or detected/default module (TextProperty_515.dll) and it will contain its own logic and model for the serialization of that version. It means going back to complex reflection and for the fronted that would be interested in displaying other fields than value this becomes also a bit more complex but in a long run it adds better overall versioned support for properties that differ between UE versions.
